File: evals/refs/nbody.zig
const std = @import("std");

fn accel(p: [][2]f64, mass: []const f64, n: usize, allocator: std.mem.Allocator) ![][2]f64 {
    const a = try allocator.alloc([2]f64, n);
    for (a) |*x| x.* = .{ 0, 0 };
    var i: usize = 0;
    while (i < n) : (i += 1) {
        var j: usize = 0;
        while (j < n) : (j += 1) {
            if (i == j) continue;
            const dx = p[j][0] - p[i][0];
            const dy = p[j][1] - p[i][1];
            const r2 = dx * dx + dy * dy;
            const r = @sqrt(r2);
            if (r > 1e-12) {
                const f = mass[j] / (r2 * r);
                a[i][0] += f * dx;
                a[i][1] += f * dy;
            }
        }
    }
    return a;
}

pub fn main(init: std.process.Init) !void {
    const arena = init.arena.allocator();
    var in_buf: [4096]u8 = undefined;
    var fr = std.Io.File.stdin().reader(init.io, &in_buf);
    const r = &fr.interface;
    var data: std.ArrayList(u8) = .empty;
    try r.appendRemainingUnlimited(arena, &data);
    var out_buf: [4096]u8 = undefined;
    var fw = std.Io.File.stdout().writer(init.io, &out_buf);
    const w = &fw.interface;
    var it = std.mem.tokenizeAny(u8, data.items, " \n\r\t");
    const t = try std.fmt.parseInt(usize, it.next().?, 10);
    var c: usize = 0;
    while (c < t) : (c += 1) {
        const n = try std.fmt.parseInt(usize, it.next().?, 10);
        const dt = try std.fmt.parseFloat(f64, it.next().?);
        const steps = try std.fmt.parseInt(usize, it.next().?, 10);
        const mass = try arena.alloc(f64, n);
        const pos = try arena.alloc([2]f64, n);
        const vel = try arena.alloc([2]f64, n);
        var i: usize = 0;
        while (i < n) : (i += 1) {
            mass[i] = try std.fmt.parseFloat(f64, it.next().?);
            pos[i][0] = try std.fmt.parseFloat(f64, it.next().?);
            pos[i][1] = try std.fmt.parseFloat(f64, it.next().?);
            vel[i][0] = try std.fmt.parseFloat(f64, it.next().?);
            vel[i][1] = try std.fmt.parseFloat(f64, it.next().?);
        }
        var a = try accel(pos, mass, n, arena);
        var s: usize = 0;
        while (s < steps) : (s += 1) {
            i = 0;
            while (i < n) : (i += 1) {
                pos[i][0] += vel[i][0] * dt + 0.5 * a[i][0] * dt * dt;
                pos[i][1] += vel[i][1] * dt + 0.5 * a[i][1] * dt * dt;
            }
            const a2 = try accel(pos, mass, n, arena);
            i = 0;
            while (i < n) : (i += 1) {
                vel[i][0] += 0.5 * (a[i][0] + a2[i][0]) * dt;
                vel[i][1] += 0.5 * (a[i][1] + a2[i][1]) * dt;
            }
            a = a2;
        }
        i = 0;
        while (i < n) : (i += 1) {
            try w.print("{d:.6} {d:.6} {d:.6} {d:.6}\n", .{ pos[i][0], pos[i][1], vel[i][0], vel[i][1] });
        }
    }
    try w.flush();
}

File: evals/refs/nbody.rs
use std::io::{self, Read, Write, BufWriter};

fn accel(p: &[[f64;2]], mass: &[f64], n: usize) -> Vec<[f64;2]> {
    let mut a = vec![[0.0f64; 2]; n];
    for i in 0..n {
        for j in 0..n {
            if i == j { continue; }
            let dx = p[j][0] - p[i][0];
            let dy = p[j][1] - p[i][1];
            let r2 = dx*dx + dy*dy;
            let r = r2.sqrt();
            if r > 1e-12 {
                let f = mass[j] / (r2 * r);
                a[i][0] += f * dx;
                a[i][1] += f * dy;
            }
        }
    }
    a
}

fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut it = s.split_ascii_whitespace();
    let t: usize = it.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let n: usize = it.next().unwrap().parse().unwrap();
        let dt: f64 = it.next().unwrap().parse().unwrap();
        let steps: usize = it.next().unwrap().parse().unwrap();
        let mut mass = vec![0.0f64; n];
        let mut pos = vec![[0.0f64;2]; n];
        let mut vel = vec![[0.0f64;2]; n];
        for i in 0..n {
            mass[i] = it.next().unwrap().parse().unwrap();
            pos[i][0] = it.next().unwrap().parse().unwrap();
            pos[i][1] = it.next().unwrap().parse().unwrap();
            vel[i][0] = it.next().unwrap().parse().unwrap();
            vel[i][1] = it.next().unwrap().parse().unwrap();
        }
        let mut a = accel(&pos, &mass, n);
        for _ in 0..steps {
            for i in 0..n {
                pos[i][0] += vel[i][0]*dt + 0.5*a[i][0]*dt*dt;
                pos[i][1] += vel[i][1]*dt + 0.5*a[i][1]*dt*dt;
            }
            let a2 = accel(&pos, &mass, n);
            for i in 0..n {
                vel[i][0] += 0.5*(a[i][0] + a2[i][0])*dt;
                vel[i][1] += 0.5*(a[i][1] + a2[i][1])*dt;
            }
            a = a2;
        }
        for i in 0..n {
            writeln!(out, "{:.6} {:.6} {:.6} {:.6}", pos[i][0], pos[i][1], vel[i][0], vel[i][1]).unwrap();
        }
    }
}
